dashboard-frontend/src/server.ts

```diff
@@ -80,6 +148,101 @@ export const makeServer = ({ environment = 'development' } = {}) => {
           },
         };
       });
+
+      this.get('/studies/:slug', ({ db }, request) => {
+        const study = (db.studies as any).findBy({ slug: request.params.slug });
+
+        return {
+          data: study,
+        };
+      });
+
+      /**
+       * Things to consider when implementing the real endpoint:
+       *  - When a Study is initially created, an associated StudyProgress must be created with default values.
+       *    As an example, when the frontend hits the real endpoint for a newly created Study, the response should
+       *    include a data array with a single StudyProgress that have default values.
+       *
+       *  - This endpoint is used by the stats card and the chart displayed in the Study page.
+       *    At the moment, we only display all-time progress (chart) or current progress (stat cards).
+       *      - So this endpoint should only return one data point per day (latest one for that day)
+       *      - This endpoint is not paginated because we need to display all available data in the chart.
+       *        If we anticipate that this will be a problem, we can limit the number of StudyProgress returned
+       *        by this endpoint to a fixed number. (first study progress and latest one will always be included)
+       *
+       *  - Soon, when we allow the users to pick a specific range of dates, we'll add the required query_parameters.
+       *    To decide yet if we want more granularity and allow the user to see hourly and minutes updates reflected in the chart.
+       */
+      this.get('/studies/:slug/progress', ({ db }, request) => {
```

Comment:
`StudyProgressResource` and `StudySegmentProgressResource` are representations that are useful for the dashboard and progress tracking. But they are not good representations for the rest of the system. In particular, they're completely redundant, `StudyProgressResource` is entirely calculable from `[]StudySegmentProgressResource`! It's just a view on that data that is of interest to the user. It's not data in and of itself and certainly shouldn't be fetched separately. 

So definitely `/progress` shouldn't need to exist.

We should have a translation layer in the frontend that translates the endpoint/backend representation into "StudyProgressResource" and "StudySegmentProgressResource" which make sense for dashboarding. 

I think the backend representation probably looks something like this: 

```javascript
interface ??? {
  id: string;
  name: string;
  datetime: number;
  current_budget: number;
  desired_percentage: number;
  expected_percentage: number;
  current_percentage: number;
  desired_participants: number | null;
  expected_participants: number;
  current_participants: number;
  current_price: number;
}
```

Where `percentageDeviationFromGoal` is a front-end only concept which is easily calculable. We can also add `deviationFromGoal` in terms of absolute participant numbers as well. And this provides a place of abstraction to add more ideas that matter for visualization but that aren't data per se. 

And actually, I suspect we will be adding more to the backend representation, like `amount_spent`, etc., but the frontend shouldn't need to worry about using it quite yet.

Reply:
I understand your point that these representations will only be used by the dashboard and due this reason, at first sight it seems that make more sense to receive the raw data from the backend and do the computation on the frontend.

Let's think what will happen, if we follow this approach and we try to represent the current UI:
- Good
  - We're still able to represent the tables

- Normal/Bad
  - We're still able to represent an overview of the current progress of an Study (stats cards ui). 
    But in order to do this, we need the current progress for all the segments of an Study, this translates into one of the following scenarios:
    - No pagination is required since we need all the data to compute the overview. All the data is refetched every X seconds leading into some overfetching.
    - We still use pagination to avoid refetching all the data every X seconds (we only fetch the latest data)
      - this complicates the frontend code a lot by introducing lots of conditional logic that involve coordination of requests

- Bad
  - The chart cannot be represented anymore. The reason why is:
    - The chart is representing data points for a specific piece of data, at a specific time. This piece of data is computed from the progress of the all the Segments of an Study at that same specific time.
      - To represent all-time data, it means fetching all the segments progress since the start from an Study

Above I listed some of the problems that I anticipate, but not all of them (due to time constraints).

I agree with you that normally it's not a good idea to reach for extra representations at the **api level** only to fulfill the needs of a single client, but I think this rule should not be set on stone and when justified such as in this case, we can make an exception.

Reply:
These are valid points, but I think I'm only convinced by them if we decide that this page doesn't need ALL the data from ALL the segments over ALL time in order to render things for the user. 

That would be bad as a requirement because It's a lot of data. But this is a data dashboard, so maybe that's ok. I doubt the transfer of the payload from the server to the client will ever be the largest bottleneck, as opposed to server calculating things, etc. 

It would be good as a requirement because it allows us to be really flexible and powerful about what we show on this dashboard, without the backend/interaction changing. We can show anything, because we have all the data, so we can calculate anything. It's all within the power of the dashboard code, because the dashboard code is the only one that cares about presentation/visualization of data.

One big question: what did we decide about the graph in the 2nd screenshot here, I can't remember: https://github.com/vlab-research/vlab/issues/3

That graph plots ALL the data. It's also the graph I use to show people how things work a lot, I think it really helps people understand what is happening, so I value it.

Reply:
Yes it's true that by returning raw data from the api and performing the computations on the dashboard, we'll need to touch the api code less often when new requirements arrive or existent one change for the dashboard which is good.

I have been thinking more about my main concern that was the over-fetching of data and maybe if that becomes a problem, we can limit the results returned to a fixed number. For example given an specific date range, the first and latest result for that range are always returned, and through some logic we pick the rest of results to return.

So yes, I think the single endpoint can work but it will be useful to discuss it thoroughly to not shoot ourselves in the foot too early. Let's talk about this in a call. 

About the second graph, before I started with the design of the app (https://github.com/vlab-research/vlab/issues/11) we decided that it will come after the other things (stat cards, area chart, tables).

But now that the other things are bootstrapped, after defining the desired outcome with you, I can start implementing it. I'll ping you in slack to talk about this.